torrent/pieces.py

```diff
@@ -0,0 +1,107 @@
+from enum import Enum
+from hashlib import sha1
+import logging
+import asyncio
+
+logger = logging.getLogger(__name__)
+
+class PieceStatus(Enum):
+    free = 0 
+    done = 2
+    in_progress = 1
+
+class Piece():
+    """The piece object reprsent one piece from pieces in the info of torrent file
+    all piece in size of piece length fro torrent file hoz me except from the last piece
+    the piece come with info hash of 20 bytes in sha1 and if the piece data the came
+    from peer in sha1 not the same like the piece hash the piece is not right"""
+    def __init__(self, piece_hash, length, piece_index):
+        self.index = piece_index
+        self.piece_hash = piece_hash
+        self.length = length
+        self.blocks = []
+        self.status = PieceStatus.free
+
+    def piece_done(self):
+        validation = self.piece_hash ==  sha1(b''.join(self.blocks)).digest()
+        if validation:
+            logger.info(f'piece {self.index} is valid')
+            self.status = PieceStatus.done
+        else:
+            logger.debug(f'piece {self.index} is not valid')
+            self.lost_piece()
+        return validation 
+
+    def lost_piece(self):
+        logger.debug(f"we lost a piece in index {self.index}")
+        self.status = PieceStatus.free
+        self.blocks = []
+
+    def add_block(self, block):
+        logger.debug(f"block add to piece {self.index}")
+        self.blocks.append(block)
+    
+    def __repr__(self):
+        return f"{self.index}:{self.status.name}"
+    
+
+class Piece_Manager():
+    """
+        the class porpuse is to manage the connection from piece to the place in
+        the memory on file. for now it only a list with data
+    """
+    def __init__(self):
+        self.pieces = []
+        self.lock = False
+
+    def add_piece(self, piece):
+        if type(piece) != Piece:
+            raise Exception(f"add piece to Piece manager failed the object appended in not piece is {self.piece}")
+        self.pieces.append(piece)
+        return True
+
+    def pieces_status(self):
+        in_progress = 0
+        free = 0
+        done = 0
+        for piece in self.pieces:
```

Comment:
i think its better to create a list of statuses `statuses = map(lambda piece: piece.status, self.pieces)` and then use list.count.
if its important to keep the ones without a valid status use a dict, but i don't think you should be checking that here. you should be checking that wherever you assign a status to a piece